### ISSUE TITLE
MIDI Learn Causing Host UI Performance Problem

### ISSUE THREAD

**Issue by williamfields** [surge-synthesizer/surge#3553]
**Bug Description:**
REAPER UI becomes laggy and less responsive when:
(1) There are multiple instances of Surge running.
(2) Every instance has at least one parameter controlled via REAPER's "MIDI Learn" functionality.
(3) MIDI controller data is sent to REAPER.

**Surge Version**
Surge Synthesizer
Plugin Host:	REAPER
System:	Intel CPU, Windows VST3, 64-bit
Build Info:	Built on 2020-12-29 at 00:27:24, using pipeline host 'fv-az80-319'
Version:	1.8.nightly.c70d689

**Reproduction Steps:**
1. Add one instance of Surge to a track in REAPER.
2. Move a control in Surge.
3. Go to: Param > Learn
4. Assign a MIDI controller and click OK.
5. Copy/paste to duplicate this instance of Surge, so there are 6 instances (or more).
6. Repeatedly send a bunch of MIDI controller data to REAPER, which includes the MIDI controller assigned above.
7. Watch the Time display on the transport. It will freeze when the MIDI controller data is sent.

**Expected Behavior:**
Host UI should not become laggy when using MIDI Learn. (This does not happen with Vital VST3.)

**Screenshots:**

Here you can see the MIDI data in the Chrome console on the right. You can also see the REAPER transport time freezing whenever the MIDI data is sent.

https://user-images.githubusercontent.com/9380730/103412876-4a852b00-4b45-11eb-8101-3d0c94ce46bd.mp4

**Computer Information (please complete the following information):**
 - OS: Windows 10
 - Host: REAPER
 - Version: 6.19

**Additional Info:**
Strangely, this also happens if you do the MIDI Learn on a *different* plugin on the same track! So, it seems that whenever MIDI is sent, if a track has a plugin with MIDI Learn enabled, REAPER is sending some kind of message to all of the plugins on that track saying "something changed on this track! update your shit!". And maybe this message is causing the problem in Surge?

**Comment by williamfields** [comment]
Two more points:

- The audio does not glitch. It does not seem to affect the audio processing.
- The Surge instances do not need to be on the same track. You could also reproduce the problem by having one Surge instance per track and duplicating the track multiple times.

**Comment by baconpaul** [comment]
Quick question: if you run with the editor windows closed does the problem resolve itself?

That will help me figure out where to debug.

**Comment by williamfields** [comment]
> Quick question: if you run with the editor windows closed does the problem resolve itself?

No. It is independent of whether or not the Surge UI is visible.

**Comment by baconpaul** [comment]
oh well goodness then it is definitely not what i thought it was

last time you attached a reaper file which was setup to show the problem. That was super duper useful. Do't suppose you could do the same again? I don't really us reaper (I have it and use it for vst3 testing but haven't used it to make music - use LPX for that)

Thanks

**Comment by williamfields** [comment]
See attached. Cutoff of filter 2 is mapped to MIDI channel 15 CC 100.

[ui glitch test3.zip](https://github.com/surge-synthesizer/surge/files/5758112/ui.glitch.test3.zip)

The tricky part is that, to reproduce the problem, you somehow need to generate a bunch of MIDI controller data to send to REAPER repeatedly.

Another interesting fact: The problem still occurs when all instances of Surge are set to "bypass". The problem does NOT occur when all instances of Surge are set to "offline". (Some background on the difference: https://forums.cockos.com/showpost.php?p=552642&postcount=2)

It's not a big deal from my perspective, but I wanted to report this because it may be symptomatic of some underlying inefficiency.

**Comment by baconpaul** [comment]
Not to be naive but would wiggling the knob on my controller mapped to 15/100 work? And is there an easy way to map to 0/40 or some such?

I wonder what it is - will be interesting to find out

**Comment by baconpaul** [comment]
Anyway I’ll peek and if it is something obvious will push a fix and if it is non obvious may wait until after we ship 18 in 10 days or so

**Comment by williamfields** [comment]
To reproduce the problem, it seems to require more controller data being sent all at once.

I'm not sure why the rest of the controller data is not ignored. Maybe by doing one "MIDI Learn" mapping, it flips a flag somewhere that means "pay attention to incoming MIDI controllers", and after that, all incoming controllers are processed/evaluated?

Thanks for your support as always!

**Comment by mkruselj** [comment]
> Maybe by doing one "MIDI Learn" mapping, it flips a flag somewhere that means "pay attention to incoming MIDI controllers", and after that, all incoming controllers are processed/evaluated?

Nope, it's not like that, last time I checked out that part of the code.

**Comment by baconpaul** [comment]
You are doing a learn in surge or a learn in reaper?

**Comment by mkruselj** [comment]
Reaper's MIDI learn was mentioned. To do that, you use the mighty VST3 right-click context menu that has the option to MIDI learn the current parameter 🙂

**Comment by baconpaul** [comment]
I don't think this will be a 18 fix but I'm tagging it 18 so I can be sure it's not a one-liner before I push it into 19

**Comment by mkruselj** [comment]
Moving this one to 1.9 milestone.

**Comment by williamfields** [comment]
I discovered that the problem only occurs if there is a large number of MIDI controller messages (like 250) coming in at the **same instant**. (Note: Each instance of Surge is only mapped to respond to 3 of them.)

If I smear the timing of the controller messages by randomly delaying each controller message between 0 and 1000 milliseconds, then the UI halting problem mostly goes away.

I am trying to figure out a way to make this reproduceable for you. Maybe I could code up a little MIDI-generator utility?

By the way, I know this is a very unusual use case for Surge, and I totally understand if you decide to ignore this entirely.

Good luck with the launch of 1.8!

**Comment by baconpaul** [comment]
Definitely not ignoring it entirely but that also explains why we couldn’t see it.
If they all come in at once I wonder what we are doing wrong! I can simulate that
But we are thinking of maybe for 19 changing quite a bit about how our plugin layer works so I’m going to pause on this issue for the next tiny bit. Lets keep it open though.

**Comment by williamfields** [comment]
This was fixed by something in 1.9. So, this can be closed.

**Comment by baconpaul** [comment]
Great! I wonder what did it. Thanks for checking!

**Comment by williamfields** [comment]
@baconpaul: I just noticed that this same issue is affecting the Clap version of Surge XT 1.3. (The VST3 version is OK.) 

Too bad we never figured out what caused it the first time.

Not sure if it is worth opening another issue. I'll probably just switch back to the VST3 version. (Without program change support, sadly. 😢)

**Comment by baconpaul** [comment]
Lemme peek with the other 1.3.1 issues but maybe after Christmas for those

**Comment by baconpaul** [comment]
OK looking at the code, i can definitely see why sending many many identical ccs at the same sample would back us up. I just need to figure out how to test it. I'm also not sure why the VST3 would act any differently than the CLAP since the code path is very similar.

Any way you happen to have a file kicking around which has the deathly-midi in it already? Something I can open in reaper and press play and see the problem would be super duper useful.

The thing is: When we get a midi cc we have to iterate over all params to find the mapping location. But we do that on the audio thread. However doing that forces refreshes which are pretty substantial.

I see some places I could optimize but really having something I can run to show the problem would help the most.

**Comment by mkruselj** [comment]
Are we not doing any sort of same value filtering? I reckon that should help some, too...

**Comment by baconpaul** [comment]
we do in one spot yes. basically if in a single block you update a param n times it only redraws it once.

but a stress case in reaper would be great. something is making the ui thread hang up and i can't see what it would be.

**Comment by williamfields** [comment]
I could probably hack together some JavaScript that generates a bunch of
controller data. I will report back later when I have some time to work on
this.

**Comment by baconpaul** [comment]
Can you give me an idea of order of magnitude of data you are sending? I have some midi generators kicking around as well.

**Comment by williamfields** [comment]
Maybe a couple hundred? I mentioned 250 in this comment: https://github.com/surge-synthesizer/surge/issues/3553#issuecomment-761180646

**Comment by Andreya-Autumn** [comment]
We'll release 1.3.1 today and this has not been addressed. So I'm bouncing this out of the milestone.

**Comment by mkruselj** [comment]
I am thinking hard about this issue, and going by lack of reports from anyone else except OP, I am inclined to consider this issue is not impacting people severely, or at all. The mere fact that you would need to send a huge number of controller events _at the same timestamp_ makes this a niche of a niche class of issues that would only happen programmatically, rather than in actual live use by a human. So I feel like closing this issue as "no big deal, don't send that many events at the same timestamp" sort of thing.

**Comment by baconpaul** [comment]
lets just make sure our modern tools don't see us doing something stupid in the code before we close it.

**Comment by mkruselj** [comment]
@baconpaul Here's what Claude said:

<hr>

This is the part that lines up with "CLAP but not VST3," and it's concrete, not speculative. Two things in `clap_direct_paramsFlush`:

```cpp
void SurgeSynthProcessor::clap_direct_paramsFlush(const clap_input_events *in,
                                                   const clap_output_events *out) noexcept
{
    uint32_t sz = in->size(in);
    for (uint32_t i = 0; i < sz; ++i)
    {
        auto ev = in->get(in, i);
        process_clap_event(ev);
        if (ev->type == CLAP_EVENT_PARAM_VALUE)
        {
            ...
            jp->processorParam->setValue(pevt->value);
        }
    }
    if (!is_clap_processing)
    {
        surge->process();
    }
}
```

Per the CLAP spec, `params.flush()` is the mechanism a host uses to deliver parameter automation **when the plugin isn't actively streaming audio** — exactly the situation of a host writing learned-CC-driven automation outside active playback, and it's typically invoked from the host's **main thread**, not the audio thread. Two things follow directly from what's on the page:

1. **Every** `CLAP_EVENT_PARAM_VALUE` in that flush batch calls `setValue()` → `setParameter01()` synchronously, inline, on whatever thread called `paramsFlush`. With 200 simultaneous learned CCs, that's 200 synchronous `setParameter01()` calls per flush per instance — and `setParameter01` (the file you just pasted) is not cheap in the general case: it does a `ReleaseControlInterpolator`, a `set_value_f01`, and a `switch` over `ctrltype` where several branches (`ct_fxtype` spawns and deletes an `Effect`, `ct_osctype`/`ct_bool_relative_switch`/etc. touch adjacent scene state) do real work — not just a flag flip.
2. **After the loop**, if `!is_clap_processing`, it calls `surge->process()` — a **full synth DSP block** — unconditionally, once per flush call, specifically *because* the plugin isn't already processing audio. The comment even flags the author's own uncertainty about this (*"the spec is unclear whether this can be called when processing"*), which tells us this branch was written defensively without a hard guarantee about the calling thread.

If REAPER is calling `paramsFlush` from its **main/UI thread** — which is the CLAP-idiomatic thing to do when a plugin is inactive and the host just wants to sync automation state — then every flush is running a full audio-block's worth of DSP synchronously **on REAPER's UI thread**. With N Surge instances each getting hammered by ~200 simultaneous CC-driven automation writes, that's N synchronous DSP calls contending for the one shared UI thread. That's a mechanism that specifically produces "REAPER UI becomes laggy," specifically scales with instance count (shared thread bottleneck, not per-instance CPU), specifically needs CC volume (more events per flush batch, and/or more frequent flush calls), and specifically wouldn't reproduce on VST3, since JUCE's VST3 automation delivery path doesn't route through this CLAP-only function at all.

## What's still unconfirmed from this file alone

- **What thread REAPER actually calls `clap_direct_paramsFlush` from**, and how often (once per learn-write, or batched). Not visible here — CLAP spec allows either main-thread-when-inactive or audio-thread-during-process, and the code's own comment suggests the author wasn't fully certain which applies. Worth confirming with a breakpoint/log on entry to `clap_direct_paramsFlush` while reproducing.
- **Whether `is_clap_processing` is actually false during the repro** (i.e., is REAPER hitting this while transport is stopped, or does it also happen during playback via `process_clap_event` inside `clap_direct_process`, which does *not* have this `surge->process()` side-call gate).
- **`setParameter01`'s cost for the specific ctrltypes actually being CC-mapped** in the report — if the learned params are mostly `ct_fxtype`/`ct_osctype`, the `spawn_effect`/`delete` and queue-based branches make each call meaningfully heavier than a generic float set.

If you can log/breakpoint `clap_direct_paramsFlush` entry with thread ID during a repro, that single data point (main thread vs audio thread, call frequency) would confirm or kill this as the root cause pretty definitively.

<hr>

Full chat for reference: [here](https://claude.ai/share/ed8c013a-f073-4ce0-986a-81ddba15f8a3)
